mmdet3d/models/losses/paconv_correlation_loss.py

```diff
@@ -0,0 +1,108 @@
+import torch
+from torch import nn as nn
+
+from mmdet3d.ops import PAConv, PAConvCUDA
+from mmdet.models.builder import LOSSES
+
+
+def weight_correlation(conv):
+    """Calculate correlations between weight kernels in Conv's weight bank as
+    regularization loss. The cosine similarity is used as metrics.
+
+    Args:
+        conv (nn.Module): A Conv modules to be regularized.
+            Currently we only support `PAConv` and `PAConvCUDA`.
+
+    Returns:
+        torch.Tensor: Correlations between each weight kernels in weight bank.
+    """
+    assert isinstance(conv, (PAConv, PAConvCUDA))
+    kernels = conv.weight_bank  # [C_in, num_kernels * C_out]
+    in_channels = conv.in_channels
+    out_channels = conv.out_channels
+    num_kernels = conv.num_kernels
+
+    # [num_kernels, Cin * Cout]
+    flatten_kernels = kernels.view(in_channels, num_kernels, out_channels).\
+        permute(1, 0, 2).reshape(num_kernels, -1)
+    # [num_kernels, num_kernels]
+    inner_product = torch.matmul(flatten_kernels, flatten_kernels.T)
+    # [num_kernels, 1]
+    kernel_norms = torch.sum(flatten_kernels**2, dim=-1, keepdim=True)**0.5
+    # [num_kernels, num_kernels]
+    kernel_norms = torch.matmul(kernel_norms, kernel_norms.T)
+    cosine_sims = inner_product / kernel_norms
+    # take upper triangular part excluding diagonal since we only compute
+    # correlation between different kernels once
+    # the square is to ensure positive loss, refer to:
+    # https://github.com/CVMI-Lab/PAConv/blob/main/scene_seg/tool/train.py#L208
+    corr = torch.sum(torch.triu(cosine_sims, diagonal=1)**2)
+
+    return corr
+
+
+def paconv_correlation_loss(modules, reduction):
+    """Computes correlation loss of PAConv modules as regularization.
+
+    Args:
+        modules (List[nn.Module] | :obj:`generator`):
+            A list or a python generator of torch.nn.Modules.
+        reduction (str): Method to reduce losses among PAConv modules.
+            The valid reduction method are none, sum or mean.
+
+    Returns:
+        torch.Tensor: Correlation loss of weight kernels.
+    """
+    corr_loss = []
+    for module in modules:
+        if isinstance(module, (PAConv, PAConvCUDA)):
+            corr_loss.append(weight_correlation(module))
+    corr_loss = torch.stack(corr_loss)
+
+    if reduction == 'sum':
+        corr_loss = torch.sum(corr_loss)
+    elif reduction == 'mean':
+        corr_loss = torch.mean(corr_loss)
+    elif reduction == 'none':
+        pass
+    else:
+        raise NotImplementedError
+
+    return corr_loss
+
+
+@LOSSES.register_module()
+class PAConvCorrelationLoss(nn.Module):
```

Comment:
if we use PAConvCorrelationLoss here, we should consider change `loss_regularization` to `loss_correlation`.

Reply:
I agree that we may need to change the name of `loss_regularization`. But don't you think `loss_correlation` is a little bit ad-hoc. I don't think any other paper will use a "correlation" loss. What about changing `PAConvCorrelationLoss` to `PAConvRegularizationLoss` so that we can retain the interface `loss_regularization`, which may be used by other segmentors in the future?

Reply:
rename to `PAConvRegularizationLoss`